Add email connector functionality

dinuish94 opened the pull request:
## Purpose
> Adds the implementation of the email connector which following operations to manipulate emails using IMAP, POP3 and SMTP protocols.

### Operations
- List emails 
- Send emails
- Mark email as read
- Mark email as deleted
- Delete email
- Expunge folder

## Goals
> Provide the ability to retrieve and transform data in emails for mediation purposes

## Approach
> The implementation built on top of javax.mail library where sending of emails is supported by SMTP protocol, listing of emails is supported by both POP3 and IMAP protocols and other operations which manipulate emails are supported by IMAP protocol.

> Furthermore, it uses an in-built connection pool which is configurable for connection pooling.

dinuish94 committed:
Add email connector files

Add initial project files related to email connector.

abeykoon commented on src/main/java/org/wso2/carbon/connector/connection/EmailConnection.java:
```diff
@@ -0,0 +1,144 @@
+package org.wso2.carbon.connector.connection;
```
Shall we add missing licence headers to all the files in the connector project?

dinuish94 replied:
Fixed

abeykoon commented on src/main/java/org/wso2/carbon/connector/connection/EmailConnection.java:
```diff
@@ -0,0 +1,144 @@
+                    replaceWithWhitespace(connectionConfiguration.getCipherSuites()));
+        }
+
+        if (connectionConfiguration.getSslProtocols() != null) {
```
Aren't these properties mandatory for creating a secured connection? If they are we need to fail the connector deployment by throwing error code "CONFIG_ERROR"?

dinuish94 replied:
These are not mandatory parameters. It is only required to specify whether to use SSL or TLS which is configured using a template parameter requireTLS. The default will be false. Other parameters are optional.

abeykoon replied:
How do we know SSL or TLS is enabled? If that is the case, these become mandatory, aren't they?

dinuish94 replied:
Based on the protocol Email Server supports, the user is expected to configure the 'requireTLS' template parameter in 'init'. Java Mail will use the default cipher suites, protocols etc. that it supports if this is not configured. These configurations are only required if it is required to configure a configuration such as a ciphersuite, which is not listed in the default list of cipher suites.

abeykoon replied:
+1. We will capture these details in the documentation.

abeykoon commented on src/main/java/org/wso2/carbon/connector/connection/EmailConnection.java:
```diff
@@ -0,0 +1,144 @@
+     * @param configString Comma separated string
+     * @return Whitespace separated string
+     */
+    private String replaceWithWhitespace(String configString) {
```
Why do we need this functionality?

dinuish94 replied:
JavaMail API expects a whitespace-separated string (https://javaee.github.io/javamail/docs/api/com/sun/mail/smtp/package-summary.html). However, since this is not a common practice, IMO it is better we let the user configure a comma-separated string on our end and then later convert it to whitespaces. WDYT?

abeykoon replied:
Yes - +1 for the approach.

abeykoon commented on src/main/java/org/wso2/carbon/connector/connection/EmailConnectionFactory.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    public boolean validateObject(Object connection) {
+
+        return ((MailBoxConnection) connection).isConnected();
```
Can a connection become malfunctioning even when it is connected? When we use the connection, if an exception is thrown related to connection better evict it and create a new one. I think the application can invalidate a connection at any time.

dinuish94 replied:
We have enabled `testOnBorrow` property, hence before returning an object this method is invoked and if the connection is stale, it will destroy the connection and fetch a new one.

abeykoon commented on src/main/java/org/wso2/carbon/connector/connection/EmailConnectionManager.java:
```diff
@@ -0,0 +1,161 @@
+
+    private static final Log log = LogFactory.getLog(EmailConnectionManager.class);
+
+    private Map<String, EmailConnection> connectionMap;
```
Let's mention what are stored here as key value pairs. It will make the reader easy to understand.

dinuish94 replied:
Fixed

abeykoon commented on src/main/java/org/wso2/carbon/connector/connection/EmailConnectionManager.java:
```diff
@@ -0,0 +1,161 @@
+    private static EmailConnectionManager manager;
+
+    private EmailConnectionManager(){
+        this.connectionMap = Collections.synchronizedMap(new HashMap<>());
```
Is there a requirement to have synchronized map? Cannot we use concurrentHashMap? https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html

dinuish94 replied:
Fixed

dinuish94 committed:
Add connectionType to connection configuration

Change protocol parameter name to 'connectionType' to generalize the connector configuration

dinuish94 committed:
Handle special properties

Handle special properties when explicitly setting the payload

dinuish94 committed:
Refactor code

dinuish94 committed:
Add UI schema

Add UI schema for Integration Studio

abeykoon commented on src/main/java/org/wso2/carbon/connector/connection/EmailConnectionManager.java:
```diff
@@ -0,0 +1,163 @@
+     *
+     * @return EmailConnectionManager instance
+     */
+    public static synchronized EmailConnectionManager getEmailConnectionManager(){
```
We better think again on making this method  synchronized. Will this make concurrent connection requests to loose overall concurrent processing?

dinuish94 replied:
Using the synchronized block is identified to have several issues as per https://rules.sonarsource.com/java/tag/multi-threading/RSPEC-2168 and this is the suggested as the compliant solution as the newer JVMs is said to have improved the performance of `synchronized`.

abeykoon commented on src/main/java/org/wso2/carbon/connector/connection/EmailConnectionManager.java:
```diff
@@ -0,0 +1,163 @@
+            // connections.
+            EmailConnectionPool pool = new EmailConnectionPool(new EmailConnectionFactory(connectionConfiguration),
+                    connectionConfiguration);
+            addConnectionPool(connectionConfiguration.getConnectionName(), pool);
```
Are we creating all connections to fill the pool here? Or are we going to create them on demand?

dinuish94 replied:
They will be created on demand when we invoke `borrowObject`.

abeykoon commented on src/main/java/org/wso2/carbon/connector/connection/EmailConnectionManager.java:
```diff
@@ -0,0 +1,163 @@
+     * Clears connection pools
+     *
+     */
+    public void clearConnectionPools(){
```
Shall we say `shutdown` connection pool?

dinuish94 replied:
Fixed

abeykoon commented on src/main/java/org/wso2/carbon/connector/connection/EmailConnectionPool.java:
```diff
@@ -0,0 +1,115 @@
+/**
+ * Connection Pool Implementation for Email Connections
+ */
+public class EmailConnectionPool extends GenericObjectPool {
```
Is it possible to move connection pool related classes to a different package? Then in the long run we can evaluate the possibility to move the implementation to the connector core with required interfaces.

dinuish94 replied:
Moved to connector core in https://github.com/wso2/carbon-mediation/pull/1425

abeykoon commented on src/main/java/org/wso2/carbon/connector/connection/MailBoxConnection.java:
```diff
@@ -0,0 +1,142 @@
+            try {
+                store.close();
+            } catch (Exception e) {
+                log.error(format("Error closing store when disconnecting. %s", e.getMessage()), e);
```
No need of e.getMessage(). It will be captured by logging exception.

dinuish94 replied:
Fixed

abeykoon commented on src/main/java/org/wso2/carbon/connector/connection/MailBoxConnection.java:
```diff
@@ -0,0 +1,142 @@
+     * @param openMode      open the folder in READ_ONLY or READ_WRITE mode
+     * @return the opened Folder
+     */
+    public synchronized Folder getFolder(String mailBoxFolder, int openMode) throws EmailConnectionException {
```
Do we need to synchronize this?

dinuish94 replied:
Fixed. No need to use synchronized

abeykoon commented on src/main/java/org/wso2/carbon/connector/operations/EmailDelete.java:
```diff
@@ -0,0 +1,79 @@
+            ResponseHandler.setErrorsInMessage(messageContext, Error.RESPONSE_GENERATION);
+            handleException(format(errorString, folder, e.getMessage()), e, messageContext);
+        } finally {
+            if (pool != null) {
```
Do we need this null check?

dinuish94 replied:
pool may become null if an exception is thrown.

abeykoon commented on src/main/java/org/wso2/carbon/connector/operations/EmailSend.java:
```diff
@@ -0,0 +1,115 @@
+                        .withAttachments(attachments)
+                        .build();
+                Transport.send(message);
+                log.debug("Email was sent successfully...");
```
Shall we add some unique field of message to identify which message was sent successfully? Otherwise this log does not serve the purpose.

dinuish94 replied:
Added subject and 'to' recipient. We do not have any unique identifier for the message assigned at this point

abeykoon commented on src/main/java/org/wso2/carbon/connector/operations/list/EmailGetAttachment.java:
```diff
@@ -0,0 +1,112 @@
+        } else {
+            ResponseHandler.setErrorsInMessage(messageContext, Error.INVALID_CONFIGURATION);
+            handleException(format("%s No emails retrieved. " +
+                    "Email list operation must be invoked first to retrieve emails.", ERROR), messageContext);
```
+1

abeykoon commented on src/main/java/org/wso2/carbon/connector/operations/list/EmailGetAttachment.java:
```diff
@@ -0,0 +1,112 @@
+        }
+        try {
+            EmailMessage emailMessage = EmailOperationUtils.getEmail(emailMessages, emailIndex);
+            if (emailMessage != null) {
```
Should we need this null check here?

dinuish94 replied:
Fixed

abeykoon commented on src/main/java/org/wso2/carbon/connector/operations/list/EmailGetBody.java:
```diff
@@ -0,0 +1,89 @@
+        try {
+            if (emailIndex != null && emailMessages != null) {
+                EmailMessage emailMessage = EmailOperationUtils.getEmail(emailMessages, emailIndex);
+                if (emailMessage != null) {
```
Do we need this null check here?

dinuish94 replied:
Fixed

abeykoon commented on src/main/java/org/wso2/carbon/connector/pojo/ConnectionConfiguration.java:
```diff
@@ -0,0 +1,290 @@
+    public void setPort(String port) throws InvalidConfigurationException {
+        if (port == null){
+            throw new InvalidConfigurationException("Mandatory parameter 'port' is not set.");
+        }
```
Shall we check if the passed port resolves to a number?

dinuish94 replied:
Fixed

abeykoon commented on src/main/java/org/wso2/carbon/connector/utils/ContentBuilder.java:
```diff
@@ -0,0 +1,191 @@
+/**
+ * Utils for building content according to their content type to the format is understandable by the mediation engine
+ */
+public final class ContentBuilder {
```
Shall we consider moving this Util to connector core so that we can reuse it in other connectors as well?

dinuish94 replied:
Moved

abeykoon commented on src/main/java/org/wso2/carbon/connector/utils/EmailAttachmentDataSource.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Email Attachment Data Source
+ */
+public class EmailAttachmentDataSource implements DataSource {
```
Shall we enhance the description stating why we need to extend the `DataSource `?

dinuish94 replied:
Fixed

abeykoon commented on src/main/java/org/wso2/carbon/connector/utils/EmailConstants.java:
```diff
@@ -0,0 +1,95 @@
+    public static final String DEFAULT_FOLDER = "INBOX";
+    public static final int DEFAULT_OFFSET = 0;
+    //TODO: Should this be infinite?
+    public static final int DEFAULT_LIMIT = 10;
```
If limit is not specified, shall we set it to -1 and retrieve all the emails in the box?

dinuish94 replied:
Fixed

abeykoon commented on src/main/java/org/wso2/carbon/connector/utils/EmailParser.java:
```diff
@@ -0,0 +1,115 @@
+     * @param message Message to be parsed
+     * @throws EmailParsingException if failed to parse the message
+     */
+    private static EmailMessage parseMessage(MimeMessage message)
```
Shall we make this util a constructor for EmailMessage?

dinuish94 replied:
Fixed

abeykoon commented on src/main/java/org/wso2/carbon/connector/utils/EmailParser.java:
```diff
@@ -0,0 +1,115 @@
+     * @return Parsed messages
+     * @throws EmailParsingException if failed to parse content
+     */
+    public static List<EmailMessage> parseMessageList(List<Message> messages) throws EmailParsingException {
```
Is this method used multiple times? If not, we can make this a private method of the referred class.

dinuish94 replied:
moved to the class level

abeykoon commented on src/main/java/org/wso2/carbon/connector/utils/MessageBuilder.java:
```diff
@@ -0,0 +1,295 @@
+import static javax.mail.Part.INLINE;
+
+/**
+ * Builds a new message
```
Builds a new MIME message to be sent as an email.

dinuish94 replied:
Fixed

abeykoon commented on src/main/java/org/wso2/carbon/connector/utils/MessageBuilder.java:
```diff
@@ -0,0 +1,295 @@
+        try (InputStream fin = new FileInputStream(file)) {
+            part.setDisposition(ATTACHMENT);
+            part.setFileName(file.getName());
+            Tika tika = new Tika();
```
Shall we check if this works  with other platforms as well - like Windows 10 and Linux?

abeykoon commented on src/main/java/org/wso2/carbon/connector/utils/ResponseHandler.java:
```diff
@@ -0,0 +1,216 @@
+/**
+ * Generates responses
+ */
+public final class ResponseHandler {
```
Shall we consider moving this to connector core (setting success response)?

dinuish94 replied:
Combined this class and Content Builder methods to one util that can be used for handling responses

dinuish94 replied:
Moved

abeykoon commented on src/main/resources/functions/list.xml:
```diff
@@ -0,0 +1,52 @@
+		<property name="recent" expression="$func:recent"/>
+		<property name="offset" expression="$func:offset"/>
+		<property name="limit" expression="$func:limit"/>
+		<property name="folder" expression="$func:folder"/>
```
What about adding all parameter values as properties to message context by default in synapse?

dinuish94 replied:
Since all the templates we have in email connector use class mediators, we have to have them assign them to context properties.

dinuish94 committed:
Refactor code

dinuish94 committed:
Move common implementations to connector core

Move connection handling and payload utils to connector core

dinuish94 committed:
Refactor code

dinuish94 committed:
Add integration tests

dinuish94 committed:
Add email connector documentation

(cherry picked from commit ca1b224333143f13557b2beedf696e78062d67f7)

dinuish94 committed:
Refactor code

dinuish94 committed:
Add new icons and hide init operation

dinuish94 committed:
Update data model
